### PR TITLE
adjust resource group logic

### PR DESCRIPTION
**Pull request by seanlongcc**

### CODE REVIEW

**Comment by seanlongcc** [comment]
- discovered that test is giving errors as the az login command we are using doesnt give the permissions required for 4.12, 4.13, and 4.14.
- need to add checks for if nothing is ran at all within the storage container ones

**Comment by seanlongcc** [comment]
- az login fixed, just needed proper perms
- add edge case of, if they exclude all storage accounts, it checks nothing and doesnt execute a test, making it N/A

File: controls/azure-foundations-cis-4.1.rb [hunk 1]
```diff
@@ -67,21 +67,37 @@
...
+                 all_storage.map { |account| "#{account['ResourceGroupName']}.#{account['StorageAccountName']}" }
+               when Hash
+                 ["#{all_storage['ResourceGroupName']}.#{all_storage['StorageAccountName']}"]
+               else
+                 []
+               end
+
+  rg_sa_list.reject! { |sa| exclusions_list.include?(sa) }
+
+  only_if('N/A - No Storage Accounts found (accounts may have been manually excluded)', impact: 0) do
```
Comment (wdower) [inline controls/azure-foundations-cis-4.1.rb]: I noticed that this one has two `only_if` clauses when I checked in Heimdall. I should've said this before but InSpec rather annoyingly [only respects one `only_if` per control](https://docs.chef.io/inspec/profiles/controls/#:~:text=Only%20one%20only_if%20is%20permitted%20per%20control%20block.%20If%20multiple%20only_if%20blocks%20are%20present%2C%20only%20the%20last%20only_if%20block%20will%20be%20honored) (the last one evaluated). If you need two conditionals like this, you can use the more-traditional way of doing it:
```ruby
if <cond>
  impact 0.0
  describe 'N/A' do
    skip "N/A because reason"
  end
else
  <the rest of the control>
end
```
It's annoying, we've asked Chef about it, they don't want to change it.

Check the profile for any other controls using two only_ifs and update them.

File: controls/azure-foundations-cis-4.1.rb
Comment (wdower) [inline controls/azure-foundations-cis-4.1.rb:1]: By the way -- I meant to ask but what's up with these clauses all through the profile?
```ruby
  only_if('N/A - No Storage Accounts found', impact: 0) do
    case all_storage
    when Array
      !all_storage.empty?
    when Hash
      !all_storage.empty?
    else
      false
    end
  end
```
Why is the API call from earlier returning either of two datatypes such that we have to do it this way?

File: controls/azure-foundations-cis-4.10.rb [hunk 2]
```diff
@@ -83,7 +83,22 @@
...
+  rg_sa_list = case all_storage
+               when Array
+                 all_storage.map { |account| "#{account['ResourceGroupName']}.#{account['StorageAccountName']}" }
+               when Hash
+                 ["#{all_storage['ResourceGroupName']}.#{all_storage['StorageAccountName']}"]
+               else
+                 []
+               end
+
+  rg_sa_list.reject! { |sa| exclusions_list.include?(sa) }
```
Comment (wdower) [inline controls/azure-foundations-cis-4.10.rb:90]: Yup, this is the kind of approach I was thinking.

**Comment by wdower** [comment]
> * az login fixed, just needed proper perms
> 
>     * add edge case of, if they exclude all storage accounts, it checks nothing and doesnt execute a test, making it N/A

If we needed updated permissions on our az account to fix this make sure they're documented in the README.

File: controls/azure-foundations-cis-5.1.4.rb [hunk 3]
```diff
@@ -98,20 +116,26 @@
       server_name = server['ServerName']
       resource_group_server = server['ResourceGroupName']
 
-      script = <<-EOH
+      if resource_group_server.to_s.empty? || server_name.to_s.empty?
+        describe 'Ensure that Microsoft Entra authentication is Configured for SQL Servers' do
+          skip 'ResourceGroupName or ServerName is empty, skipping audit test'
```
Comment (wdower) [inline controls/azure-foundations-cis-5.1.4.rb]: I think your intention is just to skip this resource group if there's no server to check and then continue the loop. But sticking an actual skip statement in here will mark the entire control as a manual assessment, which is incorrect. If you go through the loop of resource groups and there aren't actually any servers to check, _then_ we would still mark this as N/A, _not_ a manual assessment.

File: controls/azure-foundations-cis-7.1.rb [hunk 4]
```diff
@@ -81,9 +108,22 @@
...
+            Rule '#{rule['name']}' failed due to matched insecure configuration:
+            - Access: #{rule['access']} (Matched Condition: #{'Access is "Allow"' if rule['access'] == 'Allow'})
+            - Direction: #{rule['direction']} (Matched Condition: #{'Direction is "Inbound"' if rule['direction'] == 'Inbound'})
+            - Protocol: #{rule['protocol']} (Matched Condition: #{'Protocol is "TCP" or "*"' if rule['protocol'] == 'TCP' || rule['protocol'] == '*'})
+            - Destination Port Range: #{rule['destinationPortRange']} (Matched Condition: #{'Destination Port Range is "3389", "*", or matches /3389/' if rule['destinationPortRange'] == '3389' || rule['destinationPortRange'] == '*' || (rule['destinationPortRange'] =~ /3389/)})
+            - Source Address Prefix: #{rule['sourceAddressPrefix']} (Matched Condition: #{'Source Address Prefix is "*", "0.0.0.0", "/0", matches /0, "internet", or "any"' if rule['sourceAddressPrefix'] == '*' || rule['sourceAddressPrefix'] == '0.0.0.0' || rule['sourceAddressPrefix'] == '/0' || rule['sourceAddressPrefix'] =~ %r{/0} || rule['sourceAddressPrefix'].downcase == 'internet' || rule['sourceAddressPrefix'].downcase == 'any'})
```
Comment (wdower) [inline controls/azure-foundations-cis-7.1.rb]: Could make this line way shorter by just doing 
```ruby
- Source Address Prefix: #{rule['sourceAddressPrefix']} (Matched Condition: #{'Source Address Prefix is "*", "0.0.0.0", "/0", matches /0, "internet", or "any"' if ['*','0.0.0.0','/0','internet','any'].includes?(rule['sourceAddressPrefix'] })
```

File: controls/azure-foundations-cis-7.1.rb [hunk 5]
```diff
@@ -58,11 +58,38 @@
...
+      rule['access'] == 'Allow' &&
+        rule['direction'] == 'Inbound' &&
+        (rule['protocol'] == 'TCP' || rule['protocol'] == '*') &&
+        (
+          rule['destinationPortRange'] == '3389' ||
+          rule['destinationPortRange'] == '*' ||
+          (rule['destinationPortRange'] =~ /3389/)
+        ) &&
+        (
+          rule['sourceAddressPrefix'] == '*' ||
```
Comment (wdower) [inline controls/azure-foundations-cis-7.1.rb:69]: Could clean this up by doing list inclusion instead of one line per possibility:
```ruby
['*','0.0.0.0','/0','internet','any'].includes?(rule['sourceAddressPrefix']
```

File: controls/azure-foundations-cis-7.2.rb [hunk 6]
```diff
@@ -55,35 +55,73 @@
...
+    rule['access'] == 'Allow' &&
+      rule['direction'] == 'Inbound' &&
+      (rule['protocol'] == 'TCP' || rule['protocol'] == '*') &&
+      (
+        rule['destinationPortRange'] == '22' ||
+        rule['destinationPortRange'] == '*' ||
+        (rule['destinationPortRange'] =~ /22/)
+      ) &&
+      (
+        rule['sourceAddressPrefix'] == '*' ||
```
Comment (wdower) [inline controls/azure-foundations-cis-7.2.rb]: Same as 7.1

File: controls/azure-foundations-cis-7.3.rb [hunk 7]
```diff
@@ -55,32 +55,72 @@
...
+  security_rules.any? do |rule|
+    rule['access'] == 'Allow' &&
+      rule['direction'] == 'Inbound' &&
+      rule['protocol'] == 'UDP' &&
+      (
+        rule['destinationPortRange'] == '*' ||
+        (rule['destinationPortRange'] =~ /53|123|161|389|1900/)
+      ) &&
+      (
+        rule['sourceAddressPrefix'] == '*' ||
```
Comment (wdower) [inline controls/azure-foundations-cis-7.3.rb]: 7.1 etc etc

**Review by wdower — CHANGES_REQUESTED** [review]

File: controls/azure-foundations-cis-4.1.rb
Comment (seanlongcc) [reply controls/azure-foundations-cis-4.1.rb:1]: We tested it with having one storage account vs having multiple and it does return a different data type. With just one, it returns a hash, but having multiple returns an array of them.

File: controls/azure-foundations-cis-4.1.rb (hunk 1 above)
Comment (seanlongcc) [reply controls/azure-foundations-cis-4.1.rb]: resolved by having one only if, then changing the second only if into the provided example block

File: controls/azure-foundations-cis-4.1.rb
Comment (seanlongcc) [reply controls/azure-foundations-cis-4.1.rb:1]: Tested when the account doesn't exist and it returns empty. Adjusted logic in 79 controls to check if this returns .empty now

File: controls/azure-foundations-cis-5.1.4.rb (hunk 3 above)
Comment (seanlongcc) [reply controls/azure-foundations-cis-5.1.4.rb]: adjusted this control, if there is no server to check,  it will pass as NA due to the only_if

**Review by seanlongcc — COMMENTED** [review]
fix grammar

File: controls/azure-foundations-cis-6.1.1.rb [hunk 8]
```diff
@@ -114,14 +107,27 @@
...
+        failure_message = 'Diagnostic settings are not configured for the subscription'
+        expect(subject).not_to eq('[]'), failure_message
+      end
+    end
+    describe 'No resources without diagnostic settings' do
+      subject { pwsh_output.stdout.strip }
+      it 'has no resources without diagnostic settings' do
+        failure_message = "The following resources do not have a diagnostic setting: #{pwsh_output.stdout.strip}"
+        expect(subject).to be_empty, failure_message
+      end
```
Comment (seanlongcc) [inline controls/azure-foundations-cis-6.1.1.rb]: fix grammar
